perf(semantic): recycle unresolved references hash maps

overlookmotel opened the pull request:
Closes #4169.

Reduce allocations while building `Semantic` by recycling hash maps used for tracking unresolved references, rather than creating a new one for every scope.

overlookmotel commented:
* **#4213** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4213?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@lucab FYI this is what I was getting at when I suggested re-using hash maps. Sorry I explained it so badly!

Dunqing reviewed (approved):
That's great. I learned a lot.

overlookmotel committed:
perf(semantic): recycle unresolved references hash maps (#4213)

Closes #4169.

Reduce allocations while building `Semantic` by recycling hash maps used for tracking unresolved references, rather than creating a new one for every scope.
